File: algorithms/[765]情侣牵手/solution.py
# topics = ["并查集"]

from typing import List


class UnionFind:
    def __init__(self, n: int):
        self.parent = [i for i in range(n)]
        self.rank = [1] * n
        self.set_num = n

    def union(self, i: int, j: int) -> None:
        x, y = self.find(i), self.find(j)

        if x is y:
            return

        if self.rank[x] <= self.rank[y]:
            self.parent[x] = y
        else:
            self.parent[y] = x

        if self.rank[x] == self.rank[y]:
            self.rank[y] += 1

        self.set_num -= 1

    def find(self, x: int) -> int:
        if self.parent[x] is not x:
            self.parent[x] = self.find(self.parent[x])

        return self.parent[x]


class Solution:
    def minSwapsCouples(self, row: List[int]) -> int:
        n = len(row)
        uf = UnionFind(n // 2)

        for i in range(0, n, 2):
            uf.union(row[i] // 2, row[i + 1] // 2)

        # 至少交换的次数 = 所有情侣的对数 - 并查集里连通分量的个数
        return n // 2 - uf.set_num

File: algorithms/[765]情侣牵手/solution.ts
// topics = ["贪心算法"]

function minSwapsCouples(row: number[]): number {
  // 记录每个位置的情侣的位置
  const map = new Map<number, number>();
  for (let [idx, ele] of row.entries()) {
    map.set(ele ^ 1, idx); // 【技巧】异或运算 偶数 += 1; 奇数 -= 1;
  }

  let res = 0;

  // 贪心: 从前往后处理，并且每次处理都保留第 k 个位置的其中一位
  // 证明: https://leetcode-cn.com/problems/couples-holding-hands/solution/liang-chong-100-de-jie-fa-bing-cha-ji-ta-26a6/
  for (let i = 1, n = row.length; i < n; i += 2) {
    if (row[i] != (row[i - 1] ^ 1)) {
      const j = map.get(row[i - 1])!;
      [row[i], row[j]] = [row[j], row[i]];
      map.set(row[i] ^ 1, i);
      map.set(row[j] ^ 1, j);
      res += 1;
    }
  }

  return res;
}
